shared/src/main/scala/async/futures.scala

```diff
@@ -261,23 +221,67 @@ object Future:
     */
-  class Promise[T]:
-    private val myFuture = new CoreFuture[T]:
-      fut =>
-      override def cancel(): Unit =
-        if setCancelled() then fut.complete(Failure(new CancellationException()))
+  trait Promise[T] extends Future[T]:
+    inline def asFuture: Future[T] = this
+
+    /** Define the result value of `future`. */
+    def complete(result: Try[T]): Unit
 
-    /** The future defined by this promise */
-    val future: Future[T] = myFuture
+  object Promise:
+    def apply[T](): Promise[T] =
+      new CoreFuture[T] with Promise[T]:
+        override def cancel(): Unit =
+          if setCancelled() then complete(Failure(new CancellationException()))
 
-    /** Define the result value of `future`. However, if `future` was cancelled in the meantime complete with a
-      * `CancellationException` failure instead.
-      */
-    def complete(result: Try[T]): Unit = myFuture.complete(result)
+        /** Define the result value of `future`. However, if `future` was cancelled in the meantime complete with a
+          * `CancellationException` failure instead.
+          */
+        override def complete(result: Try[T]): Unit = super[CoreFuture].complete(result)
   end Promise
 
+  /** The group of handlers to be used in [[withResolver]]. As a Future is completed only once, only one of
+    * resolve/reject/complete may be used and only once.
+    */
+  trait Resolver[-T]:
+    /** Complete the future with a data item successfully */
+    def resolve(item: T): Unit = complete(Success(item))
+
+    /** Complete the future with a failure */
+    def reject(exc: Throwable): Unit = complete(Failure(exc))
+
+    /** Complete the future with the result, be it Success or Failure */
+    def complete(result: Try[T]): Unit
+
+    /** Register a cancellation handler to be called when the created future is cancelled. Note that only one handler
+      * may be used.
+      */
+    def onCancel(handler: () => Unit): Unit
+  end Resolver
+
+  /** Create a promise that may be completed asynchronously using external means.
+    *
+    * The body is run synchronously on the callers thread to setup an external asynchronous operation whose
+    * success/failure it communicates using the [[Resolver]] to complete the future.
+    *
+    * If the external operation supports cancellation, the body can register one handler using [[Resolver.onCancel]].
+    */
+  def withResolver[T](body: Resolver[T] => Unit): Promise[T] =
+    val future = new CoreFuture[T] with Resolver[T] with Promise[T] {
+      @volatile var cancelHandle = () => ()
+      override def onCancel(handler: () => Unit): Unit = cancelHandle = handler
+      override def complete(result: Try[T]): Unit = super.complete(result)
+
+      override def cancel(): Unit =
+        if setCancelled() then
+          cancelHandle()
+          reject(CancellationException())
+    }
+    body(future)
+    future
+  end withResolver
```

Comment:
I don't like the fact that we return `Promise` to the user. This means that the future contained in the `Promise` can be `complete`'d from both within the `withResolver` body **and** from outside (with `Promise.complete`.

`withResolver` should still return a `Future` imo.

Reply:
Yes, that's true. I thought about those two methods as just different means to create the same thing (an explicitly resolvable Future, i.e., a Promise), but this nature should probably be only exposed to the handler here.